javaslang/src/main/java/javaslang/collection/Iterator.java

```diff
@@ -968,7 +968,8 @@ public T getNext() {
     static <T> Iterator<T> iterate(T seed, Function<? super T, ? extends T> f) {
         Objects.requireNonNull(f, "f is null");
         return new AbstractIterator<T>() {
-            T next = seed;
+            Function<? super T, ? extends T> nextFunc = (s) -> seed;
```

Comment:
I agree that iterator should be totally lazy, thanks for providing a PR.
However, it may be simpler if we implemented it without a supplier, e.g.:
```java
return new AbstractIterator<T>() {
    T current;

    @Override
    public boolean hasNext() {
        return true;
    }

    @Override
    public T getNext() {
        current = (current == null) ? seed : f.apply(current);
        return current;
    }
};
```